icu4j/main/common_tests/src/test/java/com/ibm/icu/dev/test/format/RbnfTest.java

```diff
@@ -1908,4 +1908,34 @@ public void TestNumberingSystem() {
         rbnf.setDefaultRuleSet("%ethiopic");
         assertEquals("Wrong result with Ethiopic rule set", "፻፳፫", rbnf.format(123));
     }
+
+    @Test
+    public void TestInfiniteRecursion() {
+        final String[] badRules = {
+                ">>",
+                "<<",
+                "%foo: x=%foo=",
+                "%one: x>%two>; %two: y>%one>;"
+        };
+
+        for (String badRule : badRules) {
+            RuleBasedNumberFormat rbnf = new RuleBasedNumberFormat(badRule);
+            try {
+                rbnf.format(5);
+                // we don't actually care about the result; we just want to make sure the function returns
+            } catch (IllegalStateException e) {
+                // we're supposed to get an IllegalStateException; swallow it and continue
+            }
+
+            try {
+                rbnf.parse("foo");
+                // we don't actually care about the result; we just want to make sure the function returns
```

Comment:
Please make sure that it actually returns an error for both of these. I think the format fails with a recursion limit being hit. It shouldn't succeed.